### ISSUE TITLE
new lint: unchecked arithmetic

### ISSUE THREAD

**Issue by andrewbanchich** [rust-lang/rust-clippy#14323]
### What it does

checks for usage of standard math operators on numerical types which could produce incorrect results. this would have one lint / config setting per `unchecked_*` method.

### Advantage

finds potential overflows / underflows / math bugs.

### Drawbacks

- might be noisy, but would be added as opt-in to e.g. `correctness` group
- potential for false positives when there is currently no chance for overflow / underflow based on current use. IMO this is fine, and is inline with e.g. `clippy::unwrap_used`, which doesn't mean it WILL panic.

### Example

```rust
let foo = bar + 1;
```

Could be written as:

```rust
let foo = bar.checked_add(1);
```

**Comment by y21** [comment]
There's the [`arithmetic_side_effects`](https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects) lint which sounds similar. It's not configurable at a per-operator granularity, but I don't see much value in that either. If the goal is to have clippy require you to handle overflow explicitly everywhere, then that sounds like you want it for all operators? Are there any use cases for only banning, say `*`?

**Comment by andrewbanchich** [comment]
oooh, that's perfect. i couldn't find that one. thanks @y21
